[net] Avoid possibility of NULL pointer dereference in MarkBlockAsInFlight(...)

practicalswift opened the pull request:
In the case that the branch ...

    if (itInFlight != mapBlocksInFlight.end() && itInFlight->second.first == nodeid) {

... is taken, there was prior to this commit an implicit assumption that `MarkBlockAsInFlight(...)` was being called with its fifth and optional argument (`pit`) being present (and non-`NULL`).

There are three calls to `MarkBlockAsReceived(...)` and for two of these the optional fifth argument is not supplied, which means that `pit` is being set to the default value of `NULL`.

This commit adds an assertion which makes the mentioned assumption explicit, and removes the possibility of a `NULL` pointer dereference.

theuni commented:
I think this can be a static_assert

practicalswift commented:
@theuni But `static_assert(…)` would require a constant expression (and hence not `pit != NULL`), right? :-)

theuni commented:
Hmm, right. I suppose this would have to be a templated arg overloaded with nullptr_t to work, which would obviously be silly here. Nevermind!

JeremyRubin commented:
The right change here might be to just only write to the pointer if it is given as argument... (e.g., a branch not an assert).

Will, in normal execution, the assert ever panic?

robmcl4 commented:
From how I read it the assertion will never panic, but I agree this should be a nullcheck branch. 

I see the two calls to `MarkBlockAsReceived(..)` which exclude the final param, on lines [2260](https://github.com/bitcoin/bitcoin/blob/3908fc4728059719bed0e1c7b1c8b388c2d4a8da/src/net_processing.cpp#L2260) and [3192](https://github.com/bitcoin/bitcoin/blob/3908fc4728059719bed0e1c7b1c8b388c2d4a8da/src/net_processing.cpp#L3129). Both are protected by `cs_main`. The former [will only retrieve blocks not marked in flight](https://github.com/bitcoin/bitcoin/blob/3908fc4728059719bed0e1c7b1c8b388c2d4a8da/src/net_processing.cpp#L2235) and the latter uses `FindNextBlocksToDownload(..)`, which itself [only provides blocks not marked as in flight](https://github.com/bitcoin/bitcoin/blob/3908fc4728059719bed0e1c7b1c8b388c2d4a8da/src/net_processing.cpp#L529). So the assertion's branch will never be exercised in these two cases.

practicalswift commented:
@JeremyRubin @robmcl4 Now changed from `assert` to nullcheck!

practicalswift commented:
Any changes needed before merge? :-)

paveljanik commented on src/net_processing.cpp:
```diff
@@ -338,7 +338,9 @@ bool MarkBlockAsInFlight(NodeId nodeid, const uint256& hash, const Consensus::Pa
     map<uint256, pair<NodeId, list<QueuedBlock>::iterator> >::iterator itInFlight = mapBlocksInFlight.find(hash);
     if (itInFlight != mapBlocksInFlight.end() && itInFlight->second.first == nodeid) {
-        *pit = &itInFlight->second.second;
+        if (pit != NULL) {
```
Why do you use different style then the style used just before `return true`?

practicalswift replied:
@paveljanik Are you referring to the braces? `if (…) …` vs. `if (…) { … }`?

paveljanik replied:
`if (pit)` vs `if (pit != NULL)`

practicalswift committed:
[net] Avoid possibility of NULL pointer dereference in MarkBlockAsInFlight(...)

In the case that the branch ...

    if (itInFlight != mapBlocksInFlight.end() && itInFlight->second.first == nodeid) {

... is taken, there was prior to this commit an implicit assumption that
MarkBlockAsInFlight(...) was being called with its fifth and optional
argument (pit) being present (and non-NULL).

practicalswift commented:
@paveljanik Good point. Changed to `if (pit) { … }` for consistency and pushed. Looks good? :-)

paveljanik commented:
utACK https://github.com/bitcoin/bitcoin/pull/9549/commits/95543d8747cbf7c1945ac36c36031ae40152cf2f

Sorry for the delay.

TheBlueMatt commented:
utACK 95543d8747cbf7c1945ac36c36031ae40152cf2f

sipa commented:
utACK 95543d8747cbf7c1945ac36c36031ae40152cf2f
